### PR TITLE
nixos/systemd-boot: Fix installed version regexp

### LINKED ISSUE #140686: systemd-boot-builder.py can't deal with versions of the form major.minor
### Describe the bug

`bootctl status` shows this for me:

```
Available Boot Loaders on ESP:
          ESP: /boot (/dev/disk/by-partuuid/4491941f-d624-45fe-bd5f-3da695baf5ed)
         File: └─/EFI/systemd/systemd-bootx64.efi (systemd-boot 249.4)
         File: └─/EFI/BOOT/BOOTX64.EFI (systemd-boot 249.4)
```

When I try to upgrade my system, I get the error “could not find any previously installed systemd-boot”. This is because the `systemd-boot-builder.py` script can't deal with systemd-boot versions with a minor version appended to it, since the regex is `\d+`:

https://github.com/NixOS/nixpkgs/blob/7882b7fa236d3f96836966a6bd2a4c3c3333f1a5/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L221-L229

Fixing this is not quite straightforward sadly, since the version is later used for comparison, so fixing this requires adding support for proper major/minor version comparison in this script.

### Notify maintainers

???

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.61, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.15`
 - channels(root): `"vuizvui"`
 - channels(lukas): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/vuizvui/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: boot.loader.systemd-boot
```

### PR DESCRIPTION
**Pull request by jrobsonchase**
###### Motivation for this change
The regexp was only matching numbers and not the '.', so everyone using
systemd-boot would always see `could not find any previously installed
systemd-boot` on a `nixos-rebuild`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  ~- [ ] x86_64-darwin~
  ~- [ ] aarch64-darwin~
~- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))~
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` *<- Not sure if this is relevant?*
~- [ ] Tested execution of all binary files (usually in `./result/bin/`)~
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  ~- [ ] (Package updates) Added a release notes entry if the change is major or breaking~
  ~- [ ] (Module updates) Added a release notes entry if the change is significant~
  ~- [ ] (Module addition) Added a release notes entry if adding a new NixOS module~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
Comment (SuperSandro2000) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:221]:
```suggestion
        m = re.search("^\W+File:.*/EFI/(BOOT|systemd)/.*\.efi \(systemd-boot (\d\.\d+)\)$",
```

Comment (pennae) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:221]: sounds like a bad idea, our system has:
```
File: └─/EFI/systemd/systemd-bootx64.efi (systemd-boot 247.4-2-arch)
```

it's a lustrated system (from arch), maybe matching `systemd-boot (\d+\.\d+)[^)]*\)$` is better

Comment (jrobsonchase) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:221]: @SuperSandro2000 That's not going to work - There can be multiple digits before the `.`, and I don't know that systemd's versioning scheme guarantees exactly one `.`. Seems reasonable to allow for more version levels, especially since they're just being compared with string ordering.

@pennae Good point, I had a similar one installed earlier today. I'll add the `[^)]*` :+1:

**Comment by jrobsonchase** [comment]
Changed the `0-9` to `\d` and added `[^)]+` outside of the capture group. Ready for another review pass.

I'm still on the fence about capturing vs not capturing the extra suffix like in the arch case. Could be swayed either way.

I'm also not sure that [this check](https://github.com/NixOS/nixpkgs/blob/b8ad1c3282774603042b3c7bc2aed74bb5fe1cdb/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L227) is actually correct. Should it be "always upgrade and never downgrade" or "make it match the version the module expects?" If the later, an equivalence check would be better than the string order.

**Review by sugar700 — CHANGES_REQUESTED** [review]
That doesn't seem right. Later in this code the result of this check is used for string comparison. For instance, if a current version of systemd is 249.9, and we are going to install 249.10 then the comparison will fail.

```python
>>> "249.10" > "249.9"
False
```

(to be fair, the code below was wrong previously, however it likely wouldn't cause issues until release of systemd 1000)

**Comment by jrobsonchase** [comment]
Yeah, that's true. We'd need to parse the version as either a float if we assume it'll only ever contain one `.`, or a semver parser otherwise.

> wouldn't cause issues until release of systemd 1000

That assumes that it ever actually got to that check, which it wouldn't :stuck_out_tongue: 

Is there any reason we shouldn't just capture the entire version string and ensure that it always completely matches the version from nixpkgs? It might even be prudent to add a hash to the end of our version to make sure that it's actually the nixos binary and not just one that shares the same version.

**Comment by danielfullmer** [comment]
Just to add a bit of context here since I originally wrote this code:  I did incorrectly assume the version string would always just be a series of digits, however, if it was originally installed by NixOS then it would be.  From a closer look at the code (`boot/efi/boot.c` line 28), we see that it is set to whatever `GIT_VERSION` is set to:
```c
static const char _used_ _section_(".sdmagic") magic[] = "#### LoaderInfo: systemd-boot " GIT_VERSION " ####";
```

`GIT_VERSION` is set via `src/version/version.h.in` from `@VCS_TAG@`, which can be set via meson using `-Dversion-tag`.  (This is how arch sets their custom version tag).

Another thing is that `bootctl update` itself will only update the boot loader if it detects that its version is newer than the one present.  (see `src/boot/bootctl.c`, and track how `verb_install` will indirectly call  `copy_file_with_version_check`, which calls down to `compare_version`. Earlier this year it looks like they replaced the simple version comparator with a more complicated one (`strversmp_improved`).

Since `bootctl update` will already do the version comparison for us, and duplicating that functionality in our code would be error prone, perhaps we can just always run `bootctl update`?  I don't recall my original reasoning for duplicating the version comparison in `systemd-boot-builder.py`, but if it's not needed then we should try to get rid of it.

**Comment by jrobsonchase** [comment]
> Since bootctl update will already do the version comparison for us, and duplicating that functionality in our code would be error prone, perhaps we can just always run bootctl update? I don't recall my original reasoning for duplicating the version comparison in systemd-boot-builder.py, but if it's not needed then we should try to get rid of it

That sounds like the right approach to me. At worst, the bootctl version check will get it wrong and install an older version from nixpkgs, which is arguably more correct anyway.

**Comment by danielfullmer** [comment]
Looks like this commit would solve that issue, but I don't believe it's in the current systemd used in NixOS yet:
https://github.com/systemd/systemd/commit/e5a8b4b593e7da1823f3258f7b4e7f2d370a7c3c
And we should probably also add `--graceful` to the options.

**Comment by colemickens** [comment]
This is interesting, this change has bigger ramifications:

>Ignore failure when the EFI System Partition cannot be found, when EFI variables cannot be written, or a different or newer boot loader is already installed. Currently only applies to random seed and update operations.

I think this would actually fix a case of NixOS installs failing on systems like... say the raspberry pi4, with u-boot ->efi-> `systemd-boot`. Currently that's not an option because without `--graceful`, the install fails because it can't access UEFI variables.

It seems like `graceful` might be ... too graceful?

I really don't get why `systemd` doesn't default to:
* warn and skip if the installed version is the same or older
* have a `--reinstall` flag to force reinstall
* leave `--graceful` the way it is

Seems like it would've allowed more appropriate, granular behavior.

**Comment by jrobsonchase** [comment]
> Instead of comparing if the new version is greater, I think it would then be sufficient to just check if the version strings are different

This is exactly what I suggested [here](https://github.com/NixOS/nixpkgs/pull/140046#issuecomment-931444779), though I was unaware that `bootctl` had a version check to prevent downgrades. I'm fine with either approach ¯\\\_(ツ)_/¯

**Comment by jrobsonchase** [comment]
@SuperSandro2000 Is there anything left blocking this from being merged?

**Comment by anund** [comment]
The change works :+1:. Working systemd-boot post ~248 exposes some users to https://github.com/systemd/systemd/issues/19191 (see #143847) :-1:. The documented way to rollback the bootloader is also broken #144007 :-1:

**Comment by xaverdh** [comment]
> The change works +1. Working systemd-boot post ~248 exposes some users to [systemd/systemd#19191](https://github.com/systemd/systemd/issues/19191) (see #143847). The? documented way to rollback the bootloader is also broken #144007.

Should be patch our systemd with https://github.com/systemd/systemd/commit/e98d271e57f3d0356e444b6ea2d48836ee2769b0 before merging this?

**Comment by xaverdh** [comment]
cc @flokli (re patching systemd)

**Comment by flokli** [comment]
> cc @flokli (re patching systemd)

I replied on the issue for that (https://github.com/NixOS/nixpkgs/issues/143847#issuecomment-956006957). Let's do the discussion about patching or not there.

**Review by arianvp — APPROVED** [review]

**Comment by flokli** [comment]
I just merged https://github.com/NixOS/nixpkgs/pull/144422 into staging.

Could this be rebased on top of staging as well? I'd like to avoid this making `bootctl install` work again, and potentially breaking machines affected by https://github.com/NixOS/nixpkgs/issues/143847 until the fix from staging ends up in master.

**Comment by jrobsonchase** [comment]
> Could this be rebased on top of (and target) staging as well?

Done!

**Comment by flokli** [comment]
Thanks!

**Comment by flokli** [comment]
Those should now both end in the staging-next cycle after https://github.com/NixOS/nixpkgs/pull/144730.

**Comment by Profpatsch** [comment]
WHY IS THIS PARSED ANYWAY

I’m sorry but why the heck are we parsing display output of bootctl? Is there no machine-readable output format?

**Pull request merged**
